core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java

```diff
@@ -0,0 +1,280 @@
+            ClientFactoryOption.ADDRESS_RESOLVER_GROUP.newValue(DEFAULT_ADDRESS_RESOLVER_GROUP),
+            ClientFactoryOption.HTTP2_INITIAL_CONNECTION_WINDOW_SIZE.newValue(
+                    Flags.defaultHttp2InitialConnectionWindowSize()),
+            ClientFactoryOption.HTTP2_INITIAL_STREAM_WINDOW_SIZE.newValue(
+                    Flags.defaultHttp2InitialStreamWindowSize()),
+            ClientFactoryOption.HTTP2_MAX_FRAME_SIZE.newValue(Flags.defaultHttp2MaxFrameSize()),
+            ClientFactoryOption.HTTP2_MAX_HEADER_LIST_SIZE.newValue(Flags.defaultHttp2MaxHeaderListSize()),
+            ClientFactoryOption.HTTP1_MAX_INITIAL_LINE_LENGTH.newValue(
+                    Flags.defaultHttp1MaxInitialLineLength()),
+            ClientFactoryOption.HTTP1_MAX_HEADER_SIZE.newValue(Flags.defaultHttp1MaxHeaderSize()),
+            ClientFactoryOption.HTTP1_MAX_CHUNK_SIZE.newValue(Flags.defaultHttp1MaxChunkSize()),
+            ClientFactoryOption.IDLE_TIMEOUT_MILLIS.newValue(Flags.defaultClientIdleTimeoutMillis()),
+            ClientFactoryOption.USE_HTTP2_PREFACE.newValue(Flags.defaultUseHttp2Preface()),
+            ClientFactoryOption.USE_HTTP1_PIPELINING.newValue(Flags.defaultUseHttp1Pipelining()),
+            ClientFactoryOption.CONNECTION_POOL_LISTENER.newValue(DEFAULT_CONNECTION_POOL_LISTENER),
+            ClientFactoryOption.METER_REGISTRY.newValue(Metrics.globalRegistry)
+    };
+
+    public static final ClientFactoryOptions DEFAULT = new ClientFactoryOptions(DEFAULT_OPTIONS);
+
+    /**
+     * TBW.
+     */
+    public static ClientFactoryOptions of(ClientFactoryOptionValue<?>... options) {
+        requireNonNull(options, "options");
+        if (options.length == 0) {
+            return DEFAULT;
+        }
+        return new ClientFactoryOptions(DEFAULT, options);
+    }
+
+    private static <T> ClientFactoryOptionValue<T> filterValue(ClientFactoryOptionValue<T> optionValue) {
+        requireNonNull(optionValue, "optionValue");
+
+        final ClientFactoryOption<?> option = optionValue.option();
+        final T value = optionValue.value();
+
+        if (option == ClientFactoryOption.CHANNEL_OPTIONS) {
+            @SuppressWarnings("unchecked")
+            final ClientFactoryOption<Map<ChannelOption<?>, Object>> castOption =
+                    (ClientFactoryOption<Map<ChannelOption<?>, Object>>) option;
+            @SuppressWarnings("unchecked")
+            final ClientFactoryOptionValue<T> castOptionValue =
+                    (ClientFactoryOptionValue<T>) castOption.newValue(
+                            filterChannelOptions((Map<ChannelOption<?>, Object>) value));
+            optionValue = castOptionValue;
+        }
+
+        return optionValue;
+    }
+
+    private static Map<ChannelOption<?>, Object> filterChannelOptions(
+            Map<ChannelOption<?>, Object> channelOptions) {
+
+        requireNonNull(channelOptions, "channelOptions");
```

Comment:
You need to make an immutable copy first because `channelOptions` may be mutable.